src/properties.c

```diff
@@ -0,0 +1,683 @@
+	 */
+	g_strfreev( fields );
+}
+
+/* This is called when the TileSource changes. In particular, a new VipsImage
+ * might have been loaded, or there might no image loaded. Clean up any
+ * existing input widgets and create new ones. This is the callback for the
+ * "changed" event on @p->image_window->tile_source.
+ *
+ * @tile_source		The new tile_source, which is currently held by
+ * 			p->image_window.
+ *
+ * @m	Properties *	this
+ */
+static void
+properties_tile_source_changed( TileSource *tile_source, Properties *p )
+{
+
+#ifdef DEBUG
+	puts( "properties_tile_source_changed" );
+#endif
+
+	/* If there is a new VipsImage on the tile source, use it to create
+	 * the new main_box of user input widgets.
+	 */
+	if( tile_source->image )
+		/* The properties_create_main_box function uses the VipsImage to
+		 * dynamically create a GtkGrid of user input widgets for
+		 * viewing and editing image properties. It cleans up the old
+		 * main_box and gets the image from the new TileSource ( which is
+		 * why @tile_source isn't passed as an argument here ).
+		 */
+		properties_create_main_box( p );
+}
+
+/* This function is called when ImageWindow changes. It is connected to the
+ * "changed" signal on @p->image_window.
+ *
+ * @image_window	The new ImageWindow *
+ * @m	Properties *	this
+ */
+static void
+properties_image_window_changed( ImageWindow *image_window, Properties *p )
+{
+#ifdef DEBUG
+	puts( "properties_image_window_changed" );
+#endif
+
+	g_signal_connect_object( image_window_get_tile_source( image_window ),
+			"changed", G_CALLBACK( properties_tile_source_changed ),
+			p, 0 );
+}
+
+/* Change @p->image_window, and watch it for "changed" events. Only used
+ * privately by Properties methods.
+ *
+ * @m	Properties *	this
+ * @image_window	The new ImageWindow *
+ */
+static void
+properties_set_image_window( Properties *p, ImageWindow *image_window )
+{
+#ifdef DEBUG
+	puts( "properties_set_image_window" );
+#endif
+
+	p->image_window = image_window;
+
+	g_signal_connect_object( image_window, "changed",
+		G_CALLBACK( properties_image_window_changed ),
+		p, 0 );
+}
+
+/* Apply the values currently held by the input widgets to their corresponding
+ * properties on the.
+ *
+ * @m	Properties *	this
+ *
+ */
+void
+properties_apply( Properties *p )
+{
+	VipsImage *image;
+	TileSource *tile_source;
+	GtkWidget *label_box, *input_box, *t;
+	const char *field;
+
+#ifdef DEBUG
+	puts("properties_apply");
+#endif
+
+	/* Do nothing if there is no TileSource.
+	 */
+	if( !(tile_source = image_window_get_tile_source( p->image_window )) )
+		return;
+
+	/* If there is a TileSource, get its VipsImage @image and continue.
+	 */
+	image = tile_source_get_image( tile_source );
+
+	/* Walk through the labels in the UI ( the VipsImage property names )
+	 * and corresponding input widgets.
+	 */
+	label_box = gtk_widget_get_first_child( p->left_column );
+	input_box = gtk_widget_get_first_child( p->right_column );
+	while ( label_box ) {
```

Comment:
This is fragile -- changes to widget layout could break this code.

How about copying saveoptions and having a hash to map field name to widget-to-read-value-from?

Reply:
My thought process was that binding to an empty XML tag, and expecting the user not to add children there - would be fine. That's what a listbox is after all, and I basically just have two of them here.

You can change the layout on both of my handrolled listboxes by adding <properties> tags in the .ui template file, and you can put them wherever you want in the template - they just need to have the right id's that match the calls to the BIND macro. The id's don't have to be "right_column" and "left_column". They can be "label_box_container" and "input_box_container".

Using a map felt like overkill, since the widgets I need are all in order inside their own private boxes in the UI. I don't need to waste memory on any other data structures, because the UI is already there. That was my reasoning anyway. But having a map of strings to widgets, on the Properties widget, is fine with me! Probably clearer.